lib/model.js

```diff
@@ -3522,6 +3522,115 @@ Model.bulkWrite = function(ops, options, callback) {
+ *  according to whether or not the document is new, or whether it has changes or not.
+ *
+ * `bulkSave` uses `bulkWrite` under the hood, so it's mostly useful when dealing with many documents (10K+)
+ *
+ * @param {[Document]} documents
+ *
+ */
+Model.bulkSave = function(documents) {
+  const preSavePromises = documents.map(buildPreSavePromise);
+
+  const writeOperations = this.buildBulkWriteOperations(documents);
+
+  let bulkWriteResultPromise;
+  return Promise.all(preSavePromises)
+    .then(() => bulkWriteResultPromise = this.bulkWrite(writeOperations))
+    .then(() => documents.map(buildSuccessfulWriteHandlerPromise))
+    .then(() => bulkWriteResultPromise)
+    .catch((err) => {
+      if (!get(err, 'writeErrors.length')) {
+        throw err;
+      }
+      return Promise.all(
+        documents.map((document) => {
+          const documentError = err.writeErrors.find(writeError => {
+            const writeErrorDocumentId = writeError.err.op._id || writeError.err.op.q._id;
+            return writeErrorDocumentId.toString() === document._id.toString();
+          });
+
+          if (documentError == null) {
+            return buildSuccessfulWriteHandlerPromise(document);
+          }
+        })
+      ).then(() => {
+        throw err;
+      });
+    });
+};
+
+function buildPreSavePromise(document) {
+  return new Promise((resolve, reject) => {
+    document.schema.s.hooks.execPre('save', document, (err) => {
+      if (err) {
+        reject(err);
+        return;
+      }
+      resolve();
+    });
+  });
+}
+
+function buildSuccessfulWriteHandlerPromise(document) {
+  return new Promise((resolve, reject) => {
+    handleSuccessfulWrite(document, resolve, reject);
+  });
+}
+
+function handleSuccessfulWrite(document, resolve, reject) {
+  if (document.isNew) {
+    _setIsNew(document, false);
+  }
+
+  document.$__reset();
+  document.schema.s.hooks.execPost('save', document, {}, (err) => {
+    if (err) {
+      reject(err);
+      return;
+    }
+    resolve();
+  });
+}
+
+
+Model.buildBulkWriteOperations = function buildBulkWriteOperations(documents) {
+  if (!Array.isArray(documents)) {
+    throw new Error(`bulkSave expects an array of documents to be passed, received \`${documents}\` instead`);
+  }
+
+
+  const writeOperations = documents.reduce((accumulator, document, i) => {
+    if (!(document instanceof Document)) {
+      throw new Error(`documents.${i} was not a mongoose document, documents must be an array of mongoose documents (instanceof mongoose.Document).`);
+    }
+    const validationError = document.validateSync();
+    if (validationError) {
+      throw validationError;
+    }
+
+    const changes = document.getChanges();
+    if (document.isNew) {
+      accumulator.push({
+        insertOne: { document }
+      });
+    } else if (!utils.isEmptyObject(changes)) {
+      accumulator.push({
+        updateOne: {
```

Comment:
I'd still like for :point_down: to apply versioning and custom `$__where()` as suggested below.